### PR TITLE
Replace watcher with at_exit for Windows provider

### PR DESCRIPTION
**Pull request by elyscape**
This also removes the catalog_apply_timeout parameter.

### CODE REVIEW

File: lib/puppet/provider/reboot/windows.rb [hunk 1]
```diff
@@ -66,19 +63,9 @@ def reboot
...
-    if not File.exists?(watcher)
-      raise ArgumentError, "The watcher program #{watcher} does not exist"
-    end
-
-    Puppet.debug("Launching 'ruby.exe #{watcher}'")
-    pid = Process.spawn("ruby.exe '#{watcher}' #{Process.pid} #{@resource[:catalog_apply_timeout]} '#{shutdown_cmd}'")
-    Puppet.debug("Launched process #{pid}")
+    Puppet.debug("Adding #{shutdown_cmd} to ruby's at_exit handler")
+    pid = Process.pid
+    at_exit { system shutdown_cmd if Process.pid == pid }
```
Comment (joshcooper) [inline lib/puppet/provider/reboot/windows.rb]: when would `Process.pid != pid`?

Comment (elyscape) [inline lib/puppet/provider/reboot/windows.rb]: If a call to `Kernel#fork` occurs, at_exit will run twice.  [More info in this comment](https://github.com/puppetlabs/puppetlabs-reboot/pull/57#discussion_r21408339) by @adrienthebo.

Comment (joshcooper) [inline lib/puppet/provider/reboot/windows.rb]: That makes sense for posix, but ruby on windows doesn't implement `Kernel.fork`.

Comment (elyscape) [inline lib/puppet/provider/reboot/windows.rb]: That's a good point.  I'll remove it now.

**Comment by joshcooper** [comment]
One concern I have about the `at_exit` approach is whether they are guaranteed to execute, and in what order. For example, if a different exit handler runs first, but raises an exception, will the reboot module's exit handler still run? 

There's also the case where the reboot exit handler runs first, and another exit handler runs after, but takes awhile. The most important thing is that the agent must not still be holding the lock, when the reboot happens, otherwise, the agent will be disabled, and won't run again until the lock expires.

**Comment by adrienthebo** [comment]
For some of the `at_exit` concerns:

```
#!/usr/bin/env ruby

at_exit { puts 'first' }
at_exit { puts 'second' }
at_exit { raise Exception }

Process.kill('QUIT', $$)
```

```
└> ruby at_exit.rb
at_exit.rb:5:in `block in <main>': Exception (Exception)
second
first
[1]    20995 quit       ruby at_exit.rb
```

**Comment by adrienthebo** [comment]
I believe that the only condition that will cause `at_exit` to be skipped is if `exit!` is invoked, but the only place in the Puppet code base that does that is in `Puppet::Util::Execution` if the `exec` call failed, which is the correct behavior since the `exec` call would have prevented those handlers from running if the call succeeded.

**Comment by elyscape** [comment]
Please be advised that, for the next week, I will be on vacation and will not have internet access until I return.  I went ahead and rebased this into two commits (one for the main watcher -> at_exit conversion and one for the fix for the exit code testing) in case you guys determine that it's good and clear it for merging in my absence.  If further input from my end is needed, I'll get to it after I return on the 18th.

**Comment by elyscape** [comment]
I have returned.

**Comment by elyscape** [comment]
Ping.  Any thoughts on this?

**Comment by elyscape** [comment]
Rebased onto master.

**Comment by elyscape** [comment]
Rebased again to take #66 into account.  Pinging @joshcooper, @cyberious, @ferventcoder, and @hunner for review.

**Comment by elyscape** [comment]
Rebased to take #68 into account, and README updated with the changes.

**Comment by cyberious** [comment]
@elyscape need another rebase

**Comment by elyscape** [comment]
@cyberious Done.

**Comment by elyscape** [comment]
Rebased again.

**Comment by cyberious** [comment]
@joshcooper @Iristyle @ferventcoder any concerns with this change before I merge it?

**Comment by cyberious** [comment]
```
beaker --hosts ~/bnodes/pe/windows-2008r2-64a-inverted --options-file ./acceptance/.beaker-pe.cfg --type pe
              - Test Case Summary for suite 'tests' -
       Total Suite Time: 358.86 seconds
      Average Test Time: 18.89 seconds
              Attempted: 19
                 Passed: 19
                 Failed: 0
                Errored: 0
                Skipped: 0
                Pending: 0
                  Total: 19

      - Specific Test Case Status -
        
Failed Tests Cases:
Errored Tests Cases:
Skipped Tests Cases:
Pending Tests Cases:


No tests to run for suite 'post_suite'
Cleanup: cleaning up after successful run
```

**Pull request merged**